src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java

```diff
@@ -248,17 +279,46 @@ private static void copyFile(UnixPath source,
             // set to true when file and attributes copied
             boolean complete = false;
             try {
+                int transferSize = MIN_TRANSFER_SIZE;
+                if (transferRequiresBuffer) {
+                    int ts = MIN_TRANSFER_SIZE;
+                    try {
+                        long bss = UnixFileStoreAttributes.get(source).blockSize();
+                        long bst = UnixFileStoreAttributes.get(target).blockSize();
+                        if (bss > 0 && bst > 0) {
+                            ts = (int)(bss == bst ? bss : lcm(bss, bst));
+                        }
+                        if (ts < MIN_TRANSFER_SIZE) {
+                            int factor = (MIN_TRANSFER_SIZE + ts - 1)/ts;
+                            ts *= factor;
+                        }
+                    } catch (IllegalArgumentException | UnixException ignored) {
+                    }
+                    transferSize = ts;
+                }
+
                 // transfer bytes to target file
                 try {
+                    ByteBuffer buf = null;
+                    long address = 0L;
+                    if (transferRequiresBuffer) {
+                        buf = Util.getTemporaryDirectBuffer(transferSize);
+                        address = ((DirectBuffer)buf).address();
+                    }
                     long comp = Blocker.begin();
                     try {
-                        transfer(fo, fi, addressToPollForCancel);
+                        transfer0(fo, fi, address, transferSize,
+                                  addressToPollForCancel);
                     } finally {
+                        if (buf != null) {
+                            Util.releaseTemporaryDirectBuffer(buf);
+                        }
```

Comment:
I don't think this is maintainable as is, instead I think the two cases need to moved to supporting methods so you can have something like:

```
if (transferRequiresBuffer) {
    transferWithTemporyBuffer(...);
} else {
    transferDirect(...)
}
```

Reply:
I wrote something like that before publishing this version and did not like it but now I think it might be better.